components/tikv_util/src/logger/mod.rs

```diff
@@ -75,43 +75,57 @@ where
         }
     };
 
-    let logger = if use_async {
-        let drain = Async::new(LogAndFuse(drain))
+    let (logger, guard) = if use_async {
+        let (async_log, guard) = Async::new(LogAndFuse(drain))
             .chan_size(SLOG_CHANNEL_SIZE)
             .overflow_strategy(SLOG_CHANNEL_OVERFLOW_STRATEGY)
             .thread_name(thd_name!("slogger"))
-            .build()
-            .filter_level(level)
-            .fuse();
+            .build_with_guard();
+        let drain = async_log.filter_level(level).fuse();
         let drain = SlowLogFilter {
             threshold: slow_threshold,
             inner: drain,
         };
         let filtered = drain.filter(filter).fuse();
-        slog::Logger::root(filtered, slog_o!())
+
+        (slog::Logger::root(filtered, slog_o!()), Some(guard))
     } else {
         let drain = LogAndFuse(Mutex::new(drain).filter_level(level));
         let drain = SlowLogFilter {
             threshold: slow_threshold,
             inner: drain,
         };
         let filtered = drain.filter(filter).fuse();
-        slog::Logger::root(filtered, slog_o!())
+        (slog::Logger::root(filtered, slog_o!()), None)
     };
 
-    set_global_logger(level, init_stdlog, logger)
+    set_global_logger(level, init_stdlog, logger, guard)
+}
+
+// All Drains are reference-counted by every Logger that uses them. Async drain
+// runs a worker thread and sends a termination (and flushing) message only when
+// being dropped. Because of that it's actually quite easy to have a left-over
+// reference to a Async drain, when terminating: especially on panics or similar
+// unwinding event. Typically it's caused be a leftover reference like Logger in
+// thread-local variable, global variable, or a thread that is not being joined
+// on. So use AsyncGuard to send a flush and termination message to a Async
```

Comment:
When it's sent?

Reply:
When the async guard is dropped. In panic hook, we will replace the old global logger with a new terminal logger. At that time, the old async guard will be taken and dropped.